cupyx/jit/_compile.py

```diff
@@ -0,0 +1,323 @@
+import ast
+import inspect
+
+import numpy
+
+from cupyx.jit import _types
+from cupyx.jit import _typerules
+
+
+def transpile(func, attributes, mode, in_types, ret_type):
+    """Transpile the target function
+    Args:
+        func (function): Target function.
+        attributes (list of str): Attributes of the generated CUDA function.
+        mode ('numpy' or 'jit'): The rule for typecast.
+        in_types (list of _types.TypeBase): Types of the arguments.
```

Comment:
I wonder what happens if we have constants in the middle of the function such as `2`. How do we know if its an int32 or int64? I guess that the safest way is to declare everything as int64 to avoid overflow unless there are some kind of type hints

Reply:
Current implementation has 2 types of typerule: `numpy`-mode and `cuda`-mode, and `numpy`-mode is chosen in `cupy.vectorize`. https://github.com/cupy/cupy/blob/13b5bb57c341c8dc4733b6e1c675f462de303d6a/cupyx/jit/_typerules.py#L82-L102
In `numpy`-mode, `int` is always handled as `int64` type. In `cuda`-mode, handled as `int32` only if the value is small enough.

Reply:
cool!

Reply:
This is very nice. Numba defaults to 64 bits whenever possible, causing a lot of performance issues that need to be worked around manually by users, and IIUC they have no way to improve it as it wasn't part of original design consideration. It's critical to take this into account in the early stage as done here.